### PR TITLE
[Opt] Constant folding for BitExtractStmt

### LINKED ISSUE #1278: [Opt] Let the constant folding pass support OffsetAndExtractBitsStmt
**Concisely describe the proposed feature**
If there's an `OffsetAndExtractBitsStmt` with `input` being a `ConstStmt`, we can do constant folding.

**Describe the solution you'd like (if any)**
Add a `visit` function in `constant_fold.cpp`.

**Additional comments**
We may need to modify `hash<taichi::lang::JITEvaluatorId>` to let it support the case of an `OffsetAndExtractBitsStmt`.

Or maybe we can just calculate `(input + offset) & (((1u << (bit_end - bit_begin)) - 1) << bit_begin)` without doing JIT compilation?

### PR DESCRIPTION
**Pull request by xumingkuan**
Related issue = #656 #510 close #1278

benchmark20200622

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

File: taichi/transforms/constant_fold.cpp [hunk 1]
```diff
@@ -177,11 +177,32 @@ class ConstantFold : public BasicStmtVisitor {
...
+      return;
+    if (stmt->width() != 1)
+      return;
+    int32 result;
+    if (is_signed(input->val[0].dt)) {
+      result = (input->val[0].val_int() >> stmt->bit_begin) &
+               ((1LL << (stmt->bit_end - stmt->bit_begin)) - 1);
+    } else {
+      result = (input->val[0].val_uint() >> stmt->bit_begin) &
+               ((1LL << (stmt->bit_end - stmt->bit_begin)) - 1);
```
Comment (archibate) [inline taichi/transforms/constant_fold.cpp:197]: I thought BitExtractStmt is only used for i32?

File: taichi/transforms/constant_fold.cpp [hunk 2]
```diff
@@ -177,11 +177,32 @@ class ConstantFold : public BasicStmtVisitor {
...
-      modifier.insert_before(stmt, VecStatement(std::move(evaluated)));
+      modifier.insert_before(stmt, std::move(evaluated));
       modifier.erase(stmt);
     }
   }
 
+  void visit(BitExtractStmt *stmt) override {
+    auto input = stmt->input->cast<ConstStmt>();
+    if (!input)
+      return;
```
Comment (archibate) [inline taichi/transforms/constant_fold.cpp:188]:
```suggestion
    if (!input || input->element_type() == DataType::i32)
      return;
```

File: taichi/transforms/constant_fold.cpp (hunk 1 above)
Comment (yuanming-hu) [reply taichi/transforms/constant_fold.cpp:197]: We probably need to support unsigned inputs as well :-)

File: taichi/transforms/constant_fold.cpp [hunk 3]
```diff
@@ -177,11 +177,32 @@ class ConstantFold : public BasicStmtVisitor {
...
+    int32 result;
+    if (is_signed(input->val[0].dt)) {
+      result = (input->val[0].val_int() >> stmt->bit_begin) &
+               ((1LL << (stmt->bit_end - stmt->bit_begin)) - 1);
+    } else {
+      result = (input->val[0].val_uint() >> stmt->bit_begin) &
+               ((1LL << (stmt->bit_end - stmt->bit_begin)) - 1);
+    }
+    auto result_stmt = Stmt::make<ConstStmt>(
+        LaneAttribute<TypedConstant>(TypedConstant(DataType::i32, result)));
```
Comment (yuanming-hu) [inline taichi/transforms/constant_fold.cpp:200]: Given the part before already uses `uint` and `int` which covers all integral types, here maybe we should support all integral types as well? Just change `DataType::i32` into the input data type.

Comment (archibate) [reply taichi/transforms/constant_fold.cpp:200]: Or we may need an assertion on `input_type != i8 && :input_type != i16`?

File: taichi/transforms/constant_fold.cpp (hunk 2 above)
Comment (xumingkuan) [reply taichi/transforms/constant_fold.cpp:188]: Why not do constant folding when the `input` is `i32`?

File: taichi/transforms/constant_fold.cpp (hunk 3 above)
Comment (xumingkuan) [reply taichi/transforms/constant_fold.cpp:200]: > Given the part before already uses `uint` and `int` which covers all integral types, here maybe we should support all integral types as well? Just change `DataType::i32` into the input data type.

The `typecheck` pass set `ret_type = i32` for `OffsetAndExtractBitsStmt`. Shall we also change it to the input data type?

Comment (xumingkuan) [reply taichi/transforms/constant_fold.cpp:200]: > Or we may need an assertion on `input_type != DataType::i8 && input_type != DataType::i16`?

I think this is OK -- if the input is representable by `i8`, so is the output.

**Pull request merged**